biogeochem/EDCanopyStructureMod.F90

```diff
@@ -1919,28 +1923,68 @@ subroutine update_hlm_dynamics(nsites,sites,fcolumn,bc_out)
-           ! properties of the canopy assemblage.  Defining this needs an appropriate model.
-           ! Right now z0 and d are pft level parameters.  For the time being we will just
-           ! use the 1st index until a suitable model is defined. (RGK 04-2017)
+           
+           if (currentPatch%total_canopy_area > nearzero) then
+              currentCohort => currentPatch%shortest
+              do while(associated(currentCohort))
+                 if (currentCohort%canopy_layer .eq. 1) then
+                    weight = min(1.0_r8,currentCohort%c_area/currentPatch%total_canopy_area)
+                    bc_out(s)%z0m_pa(ifp) = bc_out(s)%z0m_pa(ifp) + &
+                         EDPftvarcon_inst%z0mr(currentCohort%pft) * currentCohort%hite * weight
+                    bc_out(s)%displa_pa(ifp) = bc_out(s)%displa_pa(ifp) + &
+                         EDPftvarcon_inst%displar(currentCohort%pft) * currentCohort%hite * weight
+                 endif
+                 currentCohort => currentCohort%taller
+              end do
+
+              ! for lai, scale to total LAI + SAI in patch.  first add up all the LAI and SAI in the patch
+              total_patch_leaf_stem_area = 0._r8
+              currentCohort => currentPatch%shortest
+              do while(associated(currentCohort))
+
+                 ! mkae sure that allometries are correct
+                 call carea_allom(currentCohort%dbh,currentCohort%n,sites(s)%spread,&
+                      currentCohort%pft,currentCohort%c_area)
+
+                 currentCohort%treelai = tree_lai(currentCohort%prt%GetState(leaf_organ, all_carbon_elements),  &
+                      currentCohort%pft, currentCohort%c_area, currentCohort%n, &
+                      currentCohort%canopy_layer, currentPatch%canopy_layer_tlai,currentCohort%vcmax25top )
+
+                 currentCohort%treesai = tree_sai(currentCohort%pft, currentCohort%dbh, currentCohort%canopy_trim, &
+                      currentCohort%c_area, currentCohort%n, currentCohort%canopy_layer, &
+                      currentPatch%canopy_layer_tlai, currentCohort%treelai , &
+                      currentCohort%vcmax25top,4)  
+
+                 total_patch_leaf_stem_area = total_patch_leaf_stem_area + &
+                      (currentCohort%treelai + currentCohort%treesai) * currentCohort%c_area
+                 currentCohort => currentCohort%taller
+              end do
+
+              ! make sure there is some leaf and stem area
+              if (total_patch_leaf_stem_area > nearzero) then
+                 currentCohort => currentPatch%shortest
+                 do while(associated(currentCohort))
+                    ! weight dleaf by the relative totals of leaf and stem area
+                    weight = (currentCohort%treelai + currentCohort%treesai) * currentCohort%c_area / total_patch_leaf_stem_area
+                    bc_out(s)%dleaf_pa(ifp) = bc_out(s)%dleaf_pa(ifp) + &
```

Comment:
Why does this part use the stem area for the weighting function? Is that to avoid strange things happening in winter when there are no leaves?

Reply:
was following this suggestion here: https://github.com/NGEET/fates/issues/690#issuecomment-707862534.  but as it should in principle be the LAI rather than the LAI+SAI, and in principle if there is no LAI on the patch then that should be handled by the HLM robustly anyway, perhaps it is better to remove the LAI from the calculation entirely.  relatedly: probably we should be doing an ELAI weighting, though that probably involves new science and would need more site-level testing.

Reply:
Did you mean remove LAI, or SAI? I think @rgknox has/had a point in his comment though, that there is still drag if there are no leaves. And rb is used in canopy conductance, i.e. in the efpot line:
https://github.com/ESCOMP/CTSM/blob/6fad071dce82ab4cbf42f6cd47399bd463bd126a/src/biogeophys/CanopyFluxesMod.F90#L1137

So I think we need some sort of value for rb relevant to the stem-only canopy?

Reply:
yes, sorry, i meant remove sai. since we're only using the lai (or lai+sai) weighting for dleaf and not the other two terms, it _should_ be safe to remove it, since if there's no leaf there wouldn't be an effect of leaf diameter?

Reply:
i see, sorry, you're right. dleaf does go into rb so needs to have a valid value even if lai = 0.  so I'll leave as-is then?

Reply:
This was waiting on me to say yes I think, so yes, that's fine...

Reply:
@ckoven I notice that we calculate update %c_area here:

https://github.com/NGEET/fates/pull/735/files#diff-315c6d4746f5566a97a47acd7630dc085c4535691dd54fbcbcd2b4f1aa5f8700R1950-R1951

But we use it right before the update here: 
https://github.com/NGEET/fates/pull/735/files#diff-315c6d4746f5566a97a47acd7630dc085c4535691dd54fbcbcd2b4f1aa5f8700R1935